### PR TITLE
[Merged by Bors] - Cache validator balances and allow them to be served over the HTTP API

### PR DESCRIPTION
**Pull request by macladson**
## Issue Addressed

#3804

## Proposed Changes

- Add `total_balance` to the validator monitor and adjust the number of historical epochs which are cached. 
- Allow certain values in the cache to be served out via the HTTP API without requiring a state read.

## Usage
```
curl -X POST "http://localhost:5052/lighthouse/ui/validator_info" -d '{"indices": [0]}' -H "Content-Type: application/json" | jq
```

```
{
  "data": {
    "validators": {
      "0": {
        "info": [
          {
            "epoch": 172981,
            "total_balance": 36566388519
          },
         ...
          {
            "epoch": 172990,
            "total_balance": 36566496513
          }
        ]
      },
      "1": {
        "info": [
          {
            "epoch": 172981,
            "total_balance": 36355797968
          },
          ...
          {
            "epoch": 172990,
            "total_balance": 36355905962
          }
        ]
      }
    }
  }
}
```

## Additional Info

This requires no historical states to operate which mean it will still function on the freshly checkpoint synced node, however because of this, the values will populate each epoch (up to a maximum of 10 entries).

Another benefit of this method, is that we can easily cache any other values which would normally require a state read and serve them via the same endpoint. However, we would need be cautious about not overly increasing block processing time by caching values from complex computations.

This also caches some of the validator metrics directly, rather than pulling them from the Prometheus metrics when the API is called. This means when the validator count exceeds the individual monitor threshold, the cached values will still be available.

### CODE REVIEW

File: beacon_node/http_api/src/ui.rs [hunk 1]
```diff
@@ -71,6 +71,79 @@ pub fn get_validator_count<T: BeaconChainTypes>(
...
+    validators: HashMap<String, ValidatorInfo>,
+}
+
+pub fn get_validator_info<T: BeaconChainTypes>(
+    request_data: ValidatorInfoRequestData,
+    chain: Arc<BeaconChain<T>>,
+) -> Result<ValidatorInfoResponse, warp::Rejection> {
+    let current_epoch = chain.epoch().map_err(beacon_chain_error)?;
+
+    let epochs = current_epoch.saturating_sub(10_u64).as_u64()..=current_epoch.as_u64();
```
Comment (AgeManning) [inline beacon_node/http_api/src/ui.rs]: The `10_u64` should probably `HISTORIC_EPOCHS` right?

**Review by AgeManning — APPROVED** [review]
This looks fine to me, although would be nice to get another review from @paulhauner  or @michaelsproul as there was an issue with large validator counts at one point.

File: beacon_node/http_api/src/ui.rs [hunk 2]
```diff
@@ -71,6 +71,79 @@ pub fn get_validator_count<T: BeaconChainTypes>(
...
+#[derive(PartialEq, Serialize, Deserialize)]
+pub struct ValidatorInfoRequestData {
+    indices: Vec<u64>,
+}
+
+#[derive(PartialEq, Serialize, Deserialize)]
+pub struct ValidatorInfoValues {
+    epoch: u64,
+    total_balance: u64,
+}
```
Comment (michaelsproul) [inline beacon_node/http_api/src/ui.rs]: We should probably quote these `u64`s for JS-safety. Seeing as the current UI works with them unquoted I think we can come back for this.

File: beacon_node/beacon_chain/src/validator_monitor.rs [hunk 3]
```diff
@@ -347,12 +410,20 @@ impl<T: EthSpec> ValidatorMonitor<T> {
             if let Some(i) = monitored_validator.index {
                 monitored_validator.touch_epoch_summary(current_epoch);
 
+                let i = i as usize;
+
+                // Cache relevant validator info.
+                if let Some(balance) = state.balances().get(i) {
+                    monitored_validator.with_epoch_summary(current_epoch, |summary| {
+                        summary.register_validator_total_balance(*balance)
```
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/validator_monitor.rs:418]: It seems a bit funky that this bypasses `individual_tracking`, but I guess that's what we want. It doesn't contribute to metric bloat, so should be fine.

**Review by michaelsproul — APPROVED** [review]
LGTM, just noted a few very minor things that we can revisit in future.

**Comment by paulhauner** [comment]
I've addressed the review comments in [57e0136](https://github.com/sigp/lighthouse/pull/3863/commits/57e01369712e3addcc4e61a5a89b653b0bffb528) and merged `unstable` to bring in the latest clippy fixes ☺️

**Comment by paulhauner** [comment]
I'm going to start bors on this now in the interests of getting a clean run for a v3.5.0 release.

@macladson could you please double check [57e0136](https://github.com/sigp/lighthouse/pull/3863/commits/57e01369712e3addcc4e61a5a89b653b0bffb528) and make sure I haven't done anything you disagree with 🙏

The `HISTORIC_EPOCHS` change is trivial and I've reached out to Mav to see if the quoting is going to be an issue. We can follow-up with another PR if the UI *must* be without quotes. 

bors r+

**Comment by macladson** [comment]
Your changes look good to me!

**Pull request closed without merging**